compiler/x/codegen/SIMDTreeEvaluator.cpp

```diff
@@ -0,0 +1,237 @@
+#include "il/Node.hpp"                                // for Node, etc
+#include "il/Node_inlines.hpp"
+#include "il/Symbol.hpp"                              // for Symbol
+#include "il/SymbolReference.hpp"
+#include "il/TreeTop.hpp"                             // for TreeTop
+#include "il/TreeTop_inlines.hpp"
+#include "il/symbol/LabelSymbol.hpp"                  // for LabelSymbol
+#include "il/symbol/MethodSymbol.hpp"                 // for MethodSymbol
+#include "il/symbol/ResolvedMethodSymbol.hpp"
+#include "infra/Assert.hpp"                           // for TR_ASSERT
+#include "infra/List.hpp"                             // for List, etc
+#include "runtime/Runtime.hpp"
+#include "x/codegen/OutlinedInstructions.hpp"
+#include "x/codegen/RegisterRematerialization.hpp"
+#include "x/codegen/X86Instruction.hpp"
+#include "x/codegen/X86Ops.hpp"                       // for ::LABEL, ::JE4, etc
+#include "x/codegen/X86Register.hpp"
+
+namespace TR { class Instruction; }
+
+static TR::MemoryReference* ConvertToPatchableMemoryReference(TR::MemoryReference* mr, TR::Node* node, TR::CodeGenerator* cg)
+   {
+   if (mr->getSymbolReference().isUnresolved())
+      {
+      // The load instructions may be wider than 8-bytes (our patching window)
+      // but we won't know that for sure until after register assignment.
+      // Hence, the unresolved memory reference must be evaluated into a register first.
+      //
+      TR::Register* tempReg = cg->allocateRegister();
+      generateRegMemInstruction(LEARegMem(cg), node, tempReg, mr, cg);
+      mr = generateX86MemoryReference(tempReg, 0, cg);
+      cg->stopUsingRegister(tempReg);
+      }
+   return mr;
+   }
+
+TR::Register* OMR::X86::TreeEvaluator::SIMDRegLoadEvaluator(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Register* globalReg = node->getRegister();
+   if (!globalReg)
+      {
+      globalReg = cg->allocateRegister(TR_VRF);
+      node->setRegister(globalReg);
+      }
+   return globalReg;
+   }
+
+TR::Register* OMR::X86::TreeEvaluator::SIMDRegStoreEvaluator(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::Node* child = node->getFirstChild();
+   TR::Register* globalReg = cg->evaluate(child);
+   cg->machine()->setXMMGlobalRegister(node->getGlobalRegisterNumber() - cg->machine()->getNumGlobalGPRs(), globalReg);
+   cg->decReferenceCount(child);
+   return globalReg;
+   }
+
+TR::Register* OMR::X86::TreeEvaluator::SIMDloadEvaluator(TR::Node* node, TR::CodeGenerator* cg)
+   {
+   TR::MemoryReference* tempMR = generateX86MemoryReference(node, cg);
+   tempMR = ConvertToPatchableMemoryReference(tempMR, node, cg);
+   TR::Register* resultReg = cg->allocateRegister(TR_VRF);
+
+   TR_X86OpCodes opCode = BADIA32Op;
+   switch (node->getSize())
+      {
+      case 16:
+         opCode = MOVDQURegMem;
+         break;
+      default:
+         break;
+      }
+   TR_ASSERT(opCode != BADIA32Op, "SIMD Load Evaluator: unsupported fill size");
```

Comment:
A CG trace statement is warranted in these cases too incase someone see a problem but does not have asserts enabled...